### DISCUSSION TITLE
401 does not throw EIdHTTPProtocolException exception

### DISCUSSION THREAD

**Discussion by regs01** [IndySockets/Indy#602] [Q&A]
Is a problem, expected behavior or am I missing something?

Every 400/500 HTTP return status code throw EIdHTTPProtocolException, but not 401. 

I have to admit that I use NewOpenSSL_PR branch with OpenSSL 1.1, but i guess it's not the case.

**Comment by rlebeau** [comment]
HTTP response code 401 is used during authentication, which `TIdHTTP` handles internally.  It is not a failure condition for application code to handle, which is why it is not generally raised as an exception, unless something goes wrong during authentication that prevents `TIdHTTP` from continuing.

`TIdHTTP` will raise an exception for 401 if one of the following conditions is met:

- the `TIdHTTP.AuthRetries` value (number of authentication attempts) reaches the `TIdHTTP.MaxAuthRetries` value for any given request.
- the `TIdHTTP.Response.WWWAuthenticate` list does not match any registered authentication scheme, and the  `TIdHTTP.OnSelectAuthorization` event (if assigned) returns `AuthenticationClass=nil`.
- the `TIdHTTP.Request.Authentication` object needs new credentials, but the `TIdHTTP.OnAuthorization` event is not assigned, or returns `Handled=False`.
- the `TIdHTTP.Request.Authentication` object fails outright and can't proceed.

**Comment by regs01** [comment]
Ah, I see. Setting `TIdHTTP.MaxAuthRetries` to `0` gives me an exception.
